### PR TITLE
make some changes for kafka adapter job-level metrics

### PR DESCRIPTION
**Pull request by zliu41**
- `AvroHdfsTimePartitionedWriter` now writes recordsWritten and bytesWritten to task state, so that the job launcher can aggregate them into job-level metrics.
- Added a noop method in `AbstractJobLauncher` so that subclasses can override it.
- `JobLauncherFactory` now supports creating job launcher based on class name, so I'll be able to write a job launcher class for Kafka.

### CODE REVIEW

**Comment by zliu41** [comment]
@liyinan926 please a take a look. Thanks.

File: gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java [hunk 1]
```diff
@@ -323,6 +324,14 @@ public void launchJob(JobListener jobListener) throws JobException {
     }
   }
 
+  /**
+   * Subclasses can override this method to do whatever processing on the taskStates,
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: Use `{@link TaskState}s`.

File: gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java [hunk 2]
```diff
@@ -323,6 +324,14 @@ public void launchJob(JobListener jobListener) throws JobException {
...
+   * e.g., aggregate task-level metrics into job-level metrics.
+   */
+  protected void processTaskStates(List<TaskState> taskStates) {
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: `postProcessTaskStates` may be a better name? BTW: I will make big changes to the commit logic here to work on a per-dataset basis. I would suggest you move the method out of the `if` block.

**Comment by zliu41** [comment]
@liyinan926 updated

**Comment by zliu41** [comment]
Added some more metrics: (1) the earliest and average timestamps each writer writes. Used to measure latency; (2) the number of kafka partitions and messages that have errors. 187b89a

**Comment by zliu41** [comment]
@liyinan926 do you have more comments?

**Comment by liyinan926** [comment]
LGTM.

**Pull request merged**